// @ndustrial/contxt-sdk/lib/coordinator/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _applications = require('./applications');

var _applications2 = _interopRequireDefault(_applications);

var _consent = require('./consent');

var _consent2 = _interopRequireDefault(_consent);

var _edgeNodes = require('./edgeNodes');

var _edgeNodes2 = _interopRequireDefault(_edgeNodes);

var _organizations = require('./organizations');

var _organizations2 = _interopRequireDefault(_organizations);

var _permissions = require('./permissions');

var _permissions2 = _interopRequireDefault(_permissions);

var _roles = require('./roles');

var _roles2 = _interopRequireDefault(_roles);

var _users = require('./users');

var _users2 = _interopRequireDefault(_users);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Module that provides access to information about Contxt
 *
 * @typicalname contxtSdk.coordinator
 */
var Coordinator = function () {
  /**
   * @param {Object} sdk An instance of the SDK so the module can communicate with other modules
   * @param {Object} request An instance of the request module tied to this module's audience.
   */
  function Coordinator(sdk, request) {
    _classCallCheck(this, Coordinator);

    var baseUrl = sdk.config.audiences.coordinator.host + '/v1';

    this._baseUrl = baseUrl;
    this._request = request;
    this._sdk = sdk;

    this._accessBaseUrl = null;
    this._deployBaseUrl = null;
    this._organizationId = null;

    this.applications = new _applications2.default(sdk, request, baseUrl);
    this.consent = new _consent2.default(sdk, request, baseUrl);
    this.edgeNodes = new _edgeNodes2.default(sdk, request, baseUrl);
    this.organizations = new _organizations2.default(sdk, request, baseUrl);
    this.permissions = new _permissions2.default(sdk, request, baseUrl);
    this.roles = new _roles2.default(sdk, request, baseUrl);
    this.users = new _users2.default(sdk, request, baseUrl);
  }

  /**
   * Sets a selected oranization ID to be used in tenant based requests
   *
   * @param {string} organizationId the ID of the organization
   *
   * @example
   * contxtSdk.coordinator
   *   .setOrganizationId('36b8421a-cc4a-4204-b839-1397374fb16b');
   */


  _createClass(Coordinator, [{
    key: 'setOrganizationId',
    value: function setOrganizationId(organizationId) {
      this._organizationId = organizationId;

      this._accessBaseUrl = this._organizationId ? this._sdk.config.audiences.coordinator.host + '/access/v1/' + this._organizationId : null;

      this._deployBaseUrl = this._organizationId ? this._sdk.config.audiences.coordinator.host + '/deploy/v1/' + this._organizationId : null;

      this.applications = new _applications2.default(this._sdk, this._request, this._deployBaseUrl || this._baseUrl, this._organizationId);
      this.consent = new _consent2.default(this._sdk, this._request, this._deployBaseUrl || this._baseUrl, this._organizationId);
      this.edgeNodes = new _edgeNodes2.default(this._sdk, this._request, this._deployBaseUrl || this._baseUrl, this._organizationId);
      this.organizations = new _organizations2.default(this._sdk, this._request, this._accessBaseUrl || this._baseUrl, this._organizationId);
      this.permissions = new _permissions2.default(this._sdk, this._request, this._accessBaseUrl || this._baseUrl, this._organizationId);
      this.roles = new _roles2.default(this._sdk, this._request, this._accessBaseUrl || this._baseUrl, this._organizationId);
      this.users = new _users2.default(this._sdk, this._request, this._accessBaseUrl || this._baseUrl, this._organizationId);
    }
  }]);

  return Coordinator;
}();

exports.default = Coordinator;
