// @ndustrial/contxt-sdk/src/config/audiences.js
export default {
  bus: {
    production: {
      clientId: 'bmoJ2jhfyIP5s4IZx5Ss6xQ9lmkoDixO',
      host: 'https://bus.ndustrial.io',
      webSocket: 'wss://bus.ndustrial.io'
    },
    staging: {
      clientId: '1HD1NG1VTBtkqRt2HRRj3E3hdmqmwzoz',
      host: 'https://bus-staging.ndustrial.io',
      webSocket: 'wss://bus-staging.ndustrial.io'
    }
  },
  contxtAuth: {
    production: {
      clientId: '75wT048QcpE7ujwBJPPjr263eTHl4gEX',
      host: 'https://contxtauth.com'
    },
    staging: {
      clientId: 'XzgumXUg5U57015haylz4zaJsiQqZy4l',
      host: 'https://contxt-auth-staging.api.ndustrial.io'
    }
  },
  coordinator: {
    production: {
      clientId: '8qY2xJob1JAxhmVhIDLCNnGriTM9bct8',
      host: 'https://contxt.api.ndustrial.io'
    },
    staging: {
      clientId: '8qY2xJob1JAxhmVhIDLCNnGriTM9bct8',
      host: 'https://contxt-staging.api.ndustrial.io'
    }
  },
  events: {
    staging: {
      clientId: 'dn4MaocJFdKtsBy9sFFaTeuJWL1nt5xu',
      host: 'https://events-staging.api.ndustrial.io'
    },
    production: {
      clientId: '7jzwfE20O2XZ4aq3cO1wmk63G9GzNc8j',
      host: 'https://events.api.ndustrial.io'
    }
  },
  facilities: {
    production: {
      clientId: 'SgbCopArnGMa9PsRlCVUCVRwxocntlg0',
      host: 'https://facilities.api.ndustrial.io'
    },
    staging: {
      clientId: 'xG775XHIOZVUn84seNeHXi0Qe55YuR5w',
      host: 'https://facilities-staging.api.ndustrial.io'
    }
  },
  files: {
    production: {
      clientId: 'VZJ2MRd71dM9oX2hB1EVAEjUe6loL7pl',
      host: 'https://files.api.ndustrial.io'
    },
    staging: {
      clientId: 'SLE310LY4nDT5p06sUGPVOLlGXk5osbQ',
      host: 'https://files-staging.api.ndustrial.io'
    }
  },
  iot: {
    staging: {
      clientId: 'm35AEcxD8hf65sq04ZU7yFxqpqVkKzES',
      host: 'https://feeds-staging.api.ndustrial.io'
    },
    production: {
      clientId: 'iznTb30Sfp2Jpaf398I5DN6MyPuDCftA',
      host: 'https://feeds.api.ndustrial.io'
    }
  }
};
