// @ndustrial/contxt-sdk/esm/sessionTypes/index.js
var TYPES = {
  AUTH0_WEB_AUTH: 'auth0WebAuth',
  MACHINE_AUTH: 'machineAuth',
  PASSWORD_GRANT_AUTH: 'passwordGrantAuth'
};

function Auth0WebAuth() {
  var Auth = require('./auth0WebAuth').default;

  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return new (Function.prototype.bind.apply(Auth, [null].concat(args)))();
}

function MachineAuth() {
  var Auth = require('./machineAuth').default;

  for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }

  return new (Function.prototype.bind.apply(Auth, [null].concat(args)))();
}

function PasswordGrantAuth() {
  var Auth = require('./passwordGrantAuth').default;

  for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
    args[_key3] = arguments[_key3];
  }

  return new (Function.prototype.bind.apply(Auth, [null].concat(args)))();
}

export { Auth0WebAuth, MachineAuth, PasswordGrantAuth, TYPES };
